### PR TITLE
[clang] Recover necessary AddrSpaceCast

### LINKED ISSUE #86791: [Clang][OpenMP] Compilation error on omp threadprivate when -fopenmp-targets=amdgcn-amd-amdhsa
The following code:
```
int GlobalVar;
#pragma omp threadprivate(GlobalVar)
int main() {
    GlobalVar = 1;
}
```
https://godbolt.org/z/nzeK5b51j

On the system, the error of the compiler is:
```
Function return type does not match operand type of return inst!
  ret ptr addrspace(1) %1
 ptrfatal error: error in backend: Broken module found, compilation aborted!
clang: error: clang frontend command failed with exit code 70 (use -v to see invocation)
```
[minTest_test_task_ThrdPrivate.zip](https://github.com/llvm/llvm-project/files/14772628/minTest_test_task_ThrdPrivate.zip)

### PR DESCRIPTION
**Pull request by JOE1994**
A necessary AddrSpaceCast was wrongfully deleted in 5c91b2886f6bf400b60ca7839069839ac3980f8f . Recover the AddrSpaceCast.

This fixes #86791 .

### CODE REVIEW

File: clang/lib/CodeGen/ItaniumCXXABI.cpp [hunk 1]
```diff
@@ -3302,6 +3302,9 @@ void ItaniumCXXABI::EmitThreadLocalInitFuncs(
       CharUnits Align = CGM.getContext().getDeclAlign(VD);
       Val = Builder.CreateAlignedLoad(Var->getValueType(), Val, Align);
     }
+    if (Val->getType() != Wrapper->getReturnType()) {
+      Val = Builder.CreateAddrSpaceCast(Val, Wrapper->getReturnType());
+    }
```
Comment (s-barannikov) [inline clang/lib/CodeGen/ItaniumCXXABI.cpp:3307]: No braces around one-line `if`.

**Review by s-barannikov — COMMENTED** [review]
May it be possible that the wrapper function's return type is invalid and the cast isn't needed?

File: clang/test/OpenMP/amdgpu_threadprivate.cpp [hunk 2]
```diff
@@ -0,0 +1,11 @@
+// REQUIRES: asserts
```
Comment (arsenm) [inline clang/test/OpenMP/amdgpu_threadprivate.cpp:1]: This does not require asserts

File: clang/test/OpenMP/amdgpu_threadprivate.cpp [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -target-cpu x86-64 -disable-llvm-passes -fopenmp-targets=amdgcn-amd-amdhsa -x c++ -emit-llvm-bc %s -o %t-x86-host.bc
+// RUN: %clang_cc1 -triple amdgcn-amd-amdhsa -aux-triple x86_64-unknown-linux-gnu -target-cpu gfx906 -fopenmp -nogpulib -fopenmp-is-target-device -fopenmp-host-ir-file-path %t-x86-host.bc -x c++ %s
+
+// Don't crash with assertions build.
```
Comment (arsenm) [inline clang/test/OpenMP/amdgpu_threadprivate.cpp:6]: Needs output checks

File: clang/lib/CodeGen/ItaniumCXXABI.cpp (hunk 1 above)
Comment (arsenm) [reply clang/lib/CodeGen/ItaniumCXXABI.cpp:3307]: Don't think you need the condition, CreateAddrSpaceCast should be a no-op if the types match

File: clang/test/OpenMP/amdgpu_threadprivate.cpp (hunk 2 above)
Comment (JOE1994) [reply clang/test/OpenMP/amdgpu_threadprivate.cpp:1]: The test crashes with an error only with an `asserts` build.
Should I just that aspect in a comment, and remove the `asserts` requirement?

File: clang/test/OpenMP/amdgpu_threadprivate.cpp [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+int MyGlobVar;
+#pragma omp threadprivate(MyGlobVar)
```
Comment (shiltian) [inline clang/test/OpenMP/amdgpu_threadprivate.cpp:8]: This should not crash for sure, but I wonder is this a valid use of OpenMP? I don't know what I should expect from this semantics.

**Comment by JOE1994** [comment]
> May it be possible that the wrapper function's return type is invalid and the cast isn't needed?

LLVM-IR (from llvm without assertions, before this revision)
shows that the wrapper function returns `ptr` while `MyGlobVar` has `addrspace(1)`.

```llvm
$thread-local wrapper routine for MyGlobVar = comdat any

@MyGlobVar = external thread_local addrspace(1) global i32, align 4
@__oclc_ABI_version = weak_odr hidden local_unnamed_addr addrspace(4) constant i32 500

define weak_odr hidden noundef ptr @thread-local wrapper routine for MyGlobVar() local_unnamed_addr #0 comdat {
  %1 = tail call align 4 ptr addrspace(1) @llvm.threadlocal.address.p1(ptr addrspace(1) align 4 @MyGlobVar)
  ret ptr addrspace(1) %1
}

declare nonnull ptr addrspace(1) @llvm.threadlocal.address.p1(ptr addrspace(1) nonnull) #1

attributes #0 = { nounwind "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="gfx906" "target-features"="+16-bit-insts,+ci-insts,+dl-insts,+dot1-insts,+dot10-insts,+dot2-insts,+dot7-insts,+dpp,+gfx8-insts,+gfx9-insts,+s-memrealtime,+s-memtime-inst,+wavefrontsize64" "uniform-work-group-size"="false" }
attributes #1 = { mustprogress nocallback nofree nosync nounwind speculatable willreturn memory(none) }
```

Updating the wrapper function's return type as below also fixes the error:

```diff
--- a/clang/lib/CodeGen/ItaniumCXXABI.cpp
+++ b/clang/lib/CodeGen/ItaniumCXXABI.cpp
@@ -3078,6 +3078,10 @@ ItaniumCXXABI::getOrCreateThreadLocalWrapper(const VarDecl *VD,
       getContext().getPointerType(RetQT), FunctionArgList());

   llvm::FunctionType *FnTy = CGM.getTypes().GetFunctionType(FI);
+  // Adjust wrapper function's ret type to have matching llvm addrspace with Val.
+  if (llvm::Type *WrapperRetTy = FnTy->getReturnType();
+      WrapperRetTy->isPointerTy() && Val->getType()->isPointerTy() && WrapperRetTy != Val->getType())
+    FnTy = llvm::FunctionType::get(Val->getType(), FnTy->params(), FnTy->isVarArg());
   llvm::Function *Wrapper =
       llvm::Function::Create(FnTy, getThreadLocalWrapperLinkage(VD, CGM),
                              WrapperName.str(), &CGM.getModule());
```

**Comment by s-barannikov** [comment]
> Updating the wrapper function's return type as below also fixes the error:

I mean, why do we have to watch for type mismatches and fix them instead of creating the correct type in the first place?
There is `getPointerType(RetQT)` just above, shouldn't it be something like `getPointerType(RetQT, <correct address space>)`?
Or RetQT should have had the correct address space, but for some reason it was lost?

**Comment by JOE1994** [comment]
> I mean, why do we have to watch for type mismatches and fix them instead of creating the correct type in the first place?
There is getPointerType(RetQT) just above, shouldn't it be something like getPointerType(RetQT, <correct address space>)?
Or RetQT should have had the correct address space, but for some reason it was lost?

I totally agree with you.
I'm unfamiliar with Clang Qualtypes
and how clang QualTypes' addrspace info
match to llvm types' addrspace info,
so I shared a quick dirty fix to at least share a direction.

**Comment by s-barannikov** [comment]
> I'm unfamiliar with Clang Qualtypes

I see. Well, neither am I :) I brought this up in case someone else could suggest the preferred way to fix the issue.
I don't have objections against the current approach, but I'm not qualified to review it either.

File: clang/test/OpenMP/amdgpu_threadprivate.cpp (hunk 2 above)
Comment (arsenm) [reply clang/test/OpenMP/amdgpu_threadprivate.cpp:1]: The test should not crash under any build. It does not matter what the symptoms are today

**Comment by arsenm** [comment]
> I mean, why do we have to watch for type mismatches and fix them instead of creating the correct type in the first place? There is `getPointerType(RetQT)` just above, shouldn't it be something like `getPointerType(RetQT, <correct address space>)`? Or RetQT should have had the correct address space, but for some reason it was lost?

Correct, this is generally how things should work. There are select contexts where a cast is necessary. 

> I'm unfamiliar with Clang Qualtypes and how clang QualTypes' addrspace info match to llvm types' addrspace info, so I shared a quick dirty fix to at least share a direction.

This depends where the value was created from. Generally you just query getTargetAddressSpace

File: clang/test/OpenMP/amdgpu_threadprivate.cpp [hunk 5]
```diff
@@ -0,0 +1,17 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -target-cpu x86-64 -disable-llvm-passes -fopenmp-targets=amdgcn-amd-amdhsa -x c++ -emit-llvm-bc %s -o %t-x86-host.bc
+// RUN: %clang_cc1 -triple amdgcn-amd-amdhsa -aux-triple x86_64-unknown-linux-gnu -target-cpu gfx906 -fopenmp -nogpulib -fopenmp-is-target-device -fopenmp-host-ir-file-path %t-x86-host.bc -x c++ -emit-llvm %s -o - | FileCheck %s
```
Comment (arsenm) [inline clang/test/OpenMP/amdgpu_threadprivate.cpp:2]: Shouldn't need -x, or the host run line?

Comment (JOE1994) [reply clang/test/OpenMP/amdgpu_threadprivate.cpp]: Thank you for your feedback! I've pushed another follow-up commit.

File: clang/test/OpenMP/amdgpu_threadprivate.cpp [hunk 6]
```diff
@@ -0,0 +1,16 @@
+// RUN: %clang_cc1 -triple amdgcn-amd-amdhsa -aux-triple x86_64-unknown-linux-gnu -target-cpu gfx906 -fopenmp -nogpulib -fopenmp-is-target-device -emit-llvm %s -o - | FileCheck %s
+
+// Don't crash with assertions build.
+
+// CHECK:       @MyGlobVar = external thread_local addrspace(1) global i32, align 4
+// CHECK:       define weak_odr hidden noundef ptr @_ZTW9MyGlobVar() #0 comdat {
+// CHECK-NEXT:    %1 = call align 4 ptr addrspace(1) @llvm.threadlocal.address.p1(ptr addrspace(1) align 4 @MyGlobVar)
```
Comment (arsenm) [inline clang/test/OpenMP/amdgpu_threadprivate.cpp:7]: Do we support this intrinsic?

**Review by arsenm — APPROVED** [review]

File: clang/test/OpenMP/amdgpu_threadprivate.cpp (hunk 6 above)
Comment (JOE1994) [reply clang/test/OpenMP/amdgpu_threadprivate.cpp:7]: (For AMDGPU codegen) Doesn't seem so to me.

**Pull request merged**